In [1]:
from tensorflow.keras.datasets import mnist #Библиотека с базой Mnist
from tensorflow.keras.models import Sequential # Подлючаем класс создания модели Sequential
from tensorflow.keras.layers import Dense # Подключаем класс Dense - полносвязный слой
from tensorflow.keras.optimizers import Adam # Подключаем оптимизатор Adam
from tensorflow.keras import utils #Утилиты для to_categorical
from tensorflow.keras.preprocessing import image #Для отрисовки изображения
import numpy as np # Подключаем библиотеку numpy
import pylab # Модуль для построения графиков
from mpl_toolkits.mplot3d import Axes3D # Модуль для трехмерной графики
from google.colab import files #Для загрузки своей картинки
import matplotlib.pyplot as plt #Отрисовка изображений
from PIL import Image #Отрисовка изображений
#Отрисовывать изображения в ноутбуке, а не в консоль или файл
%matplotlib inline 

ModuleNotFoundError: No module named 'tensorflow'

In [ ]:
(xtrainOrg, ytrainOrg), (xtestOrg, ytestOrg)= mnist.load_data()

11493376/11490434 [==============================] - 0s 0us/step


In [ ]:
xTrain=xtrainOrg.reshape(60000,784)
xTest=xtestOrg.reshape(10000,784)
print(xtrainOrg.shape,xTrain.shape,xtestOrg.shape,xTest.shape)

(60000, 28, 28) (60000, 784) (10000, 28, 28) (10000, 784)


In [ ]:
xTrain = xTrain.astype('float32') 
xTrain = xTrain / 255 
xTest = xTest.astype('float32')
xTest = xTest / 255
print(xTrain.dtype,xTest.dtype)

float32 float32


In [ ]:
yTrain = utils.to_categorical(ytrainOrg, 10)
yTest = utils.to_categorical(ytestOrg, 10)

In [ ]:
mymodel=Sequential()
mymodel.add(Dense(10, input_dim=784, activation="relu"))
mymodel.add(Dense(400, activation="relu"))
mymodel.add(Dense(10, activation="softmax"))

In [ ]:
mymodel.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])
print(mymodel.summary())

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 10)                7850      
_________________________________________________________________
dense_1 (Dense)              (None, 400)               4400      
_________________________________________________________________
dense_2 (Dense)              (None, 10)                4010      
Total params: 16,260
Trainable params: 16,260
Non-trainable params: 0
_________________________________________________________________
None


In [ ]:
mymodel.fit(xTrain, yTrain, batch_size=128, epochs=15, verbose=1)

Epoch 1/15
469/469 [==============================] - 2s 2ms/step - loss: 0.9653 - accuracy: 0.6801
Epoch 2/15
469/469 [==============================] - 1s 2ms/step - loss: 0.2743 - accuracy: 0.9174
Epoch 3/15
469/469 [==============================] - 1s 2ms/step - loss: 0.2203 - accuracy: 0.9346
Epoch 4/15
469/469 [==============================] - 1s 2ms/step - loss: 0.1936 - accuracy: 0.9422
Epoch 5/15
469/469 [==============================] - 1s 2ms/step - loss: 0.1692 - accuracy: 0.9494
Epoch 6/15
469/469 [==============================] - 1s 2ms/step - loss: 0.1533 - accuracy: 0.9540
Epoch 7/15
469/469 [==============================] - 1s 2ms/step - loss: 0.1467 - accuracy: 0.9558
Epoch 8/15
469/469 [==============================] - 1s 2ms/step - loss: 0.1369 - accuracy: 0.9569
Epoch 9/15
469/469 [==============================] - 1s 2ms/step - loss: 0.1319 - accuracy: 0.9589
Epoch 10/15
469/469 [==============================] - 1s 2ms/step - loss: 0.1261 - accuracy: 0.9606

In [ ]:
ntest=10000 #число тестируемых цифр
lotx = xTest[range(ntest)]
lotxExp=np.expand_dims(lotx,axis=0)
lotpredict=mymodel.predict(lotxExp)
print('номера неправильно угаданных цифр из тествого набора',
      np.where(ytestOrg[range(ntest)]!=np.argmax(lotpredict,axis=2))[1])
print('все ли цифры угаданы верно:',(ytestOrg[range(ntest)]==np.argmax(lotpredict,axis=2)).all())# все ли цифры угаданы верно
print('доля неугаданных',np.where(ytestOrg[range(ntest)]!=
               np.argmax(lotpredict,axis=2))[1].size/ntest)# доля неугаданных

номера неправильных цифр из тествого набора [  18  104  115  149  151  247  259  290  320  321  359  445  447  449
  495  582  591  610  619  646  659  664  691  707  717  726  877  883
  936  939  951  956  965 1014 1101 1107 1112 1114 1156 1159 1166 1178
 1192 1226 1232 1242 1247 1260 1270 1319 1378 1393 1433 1494 1522 1527
 1530 1549 1553 1554 1559 1609 1614 1621 1641 1681 1737 1790 1828 1853
 1868 1901 1909 1940 1941 1952 1984 2018 2052 2053 2063 2070 2093 2098
 2104 2109 2118 2130 2135 2162 2182 2224 2232 2237 2266 2280 2358 2369
 2387 2408 2454 2462 2478 2488 2526 2534 2544 2597 2598 2648 2654 2705
 2720 2758 2770 2823 2853 2863 2877 2896 2915 2921 2927 2938 2953 2995
 3005 3030 3060 3073 3108 3132 3146 3289 3333 3336 3376 3377 3384 3388
 3422 3457 3475 3503 3520 3534 3549 3558 3559 3567 3604 3629 3662 3664
 3702 3723 3726 3742 3751 3776 3780 3808 3850 3853 3859 3869 3893 3906
 3941 3971 3985 4007 4017 4065 4078 4102 4116 4140 4154 4156 4163 4176
 4201 4205 4212 4248 4300 4317 43

In [ ]:
import pandas as pd
table=pd.DataFrame({'loss':[0.009,0.1029,0.0093,0.0063,0.2709,0.1463,0.021,9.10E-03,0.0901],
                    'accuracy': [0.997,0.9669,0.9969,0.9976,0.9222,0.9775,0.995,0.997,0.9037],
                    '1-accuracy': [0.003,0.0331,0.0031,0.0024,0.0778,0.0225,0.005,0.003,0.0963],
                    'доля ошибок в тесте': [0.0197,0.0414,0.0236,0.0163,0.0815,0.0345,0.0226,0.0257,0.3368]},
                   index=['стандартные настройки','1й слой 10 нейронов','1й слой 100 нейронов','1й слой 5000 нейронов',
                          'активация: linear','batch_size: 1','batch_size: 10','batch_size: 100','batch_size: 60000'])


table

loss  accuracy  1-accuracy  доля ошибок в тесте
стандартные настройки  0.0090    0.9970      0.0030               0.0197
1й слой 10 нейронов    0.1029    0.9669      0.0331               0.0414
1й слой 100 нейронов   0.0093    0.9969      0.0031               0.0236
1й слой 5000 нейронов  0.0063    0.9976      0.0024               0.0163
активация: linear      0.2709    0.9222      0.0778               0.0815
batch_size: 1          0.1463    0.9775      0.0225               0.0345
batch_size: 10         0.0210    0.9950      0.0050               0.0226
batch_size: 100        0.0091    0.9970      0.0030               0.0257
batch_size: 60000      0.0901    0.9037      0.0963               0.3368

Выводы.

Количество нейронов в 1-м слое значительно меньшее 100 ухудшает на 3% точность, а значительно большее 100 уже не сильно приближает точность к 100%, поэтому нет смысла так сильно увеличивать это число, если у нас ограниченные выислительные ресурсы и нам не нужна максимальная точность.

Линейная активация уменьшает точность на 7%, значит хуже подходит для данной задачи.

При длине блока 1 модель обучается очень долго и точность достигает 98% против 99,7% при стандартных настройках, т.е. такое обучение не дает какого-то выигрыша, а значит явно не рационально. При длине блока 10 и 100 точность примерно равна стандартной и различие только во времени: время понижается при увеличении длины блока. При 60000 точность уже значительно понижается при минимальном времени, т.е. здесь мы отходим от оптимума, поэтому длина в 100 примеров близка к отимальной, если нет строгих ограничений по времени.